### PR TITLE
feat(tocObj): bring up

### LINKED ISSUE #136: Feature Request: TOC
If we are going to address something like https://github.com/hexojs/hexo/issues/3287, we will have duplicated dom process code in toc helper and toc tag plugin. 

So maybe we can address a function that could parse titles from HTML string into JSON object. Then we can use this json to build either toc helper or toc tag plugin. I might name this function to `tocObj()`.

I have came up an idea of  the format:

```html
<div id="para-1">
    <h1>Paragraph 1</h1>
</div>
<h2 id="para-1-1">Paragraph 1.1</h2>
<h1 id="para-2">Paragraph 2</h1>
<h2 id="para-2-1">Paragraph 2.1</h2>
<div id="para-2-2">
    <div>
        <h2>Paragraph 2.2</h2>
    </div>
</div>
<h3 id="para-2-2-1">Paragraph 2.2.1</h3>
```

```json
[
    {
        "name": "Paragraph 1",
        "level": 1,
        "id": "para-1",
        "children": [
            {
                "name": "Paragraph 1.1",
                "level": 2,
                "id": "para-1-1",
                "children": []
            }
        ]
    },
    {
        "name": "Paragraph 2",
        "level": 1,
        "id": "para-2",
        "children": [
            {
                "name": "Paragraph 2.1",
                "level": 2,
                "id": "para-2-1",
                "children": []
            },
            {
                "name": "Paragraph 2.2",
                "level": 2,
                "id": "para-2-3",
                "children": [
                    {
                        "name": "Paragraph 2.2.1",
                        "level": 3,
                        "id": "para-2-2-1",
                        "children": []
                    }
                ]
            }
        ]
    }
]
```

or simpler:

```
[ { name: 'Paragraph 1', id: 'para-1', lvl: 1 },
  { name: 'Paragraph 2', id: 'para-2', lvl: 1 },
  { name: 'Paragraph 2.1', id: 'para-2-1', lvl: 2 },
  { name: 'Paragraph 2.2', id: 'para-2-2', lvl: 2 },
  { name: 'Paragraph 2.2.1', id: 'para-2-2-1', lvl: 3 } ]
```

### PR DESCRIPTION
**Pull request by SukkaW**
Close #136.

### CODE REVIEW

**Comment by SukkaW** [comment]
IMHO, we should use another name for this feature. Maybe `tocUtil()` or `parseToc()`?

File: README.md [hunk 1]
```diff
@@ -409,6 +410,67 @@ wordWrap('Once upon a time', {width: 1})
 // Once\nupon\na\ntime
 ```
 
+### tocObj(str, [options])
+
+Generate a table of contents in JSON format based on the given html string.
+
+Option | Description | Default
+--- | --- | ---
+`min_depth` | The minimum level of TOC | 0
```
Comment (noraj) [inline README.md]:
```diff
- `min_depth` | The minimum level of TOC | 0 
+ `min_depth` | The minimum level of TOC | 1 
```

I think I should be 1 there is no such a thing a heading of level zero.

File: lib/toc_obj.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+const getId = (ele) => {
+  const { id } = ele.attribs;
+  const { parent } = ele;
+  return id || (parent.length < 1 ? null : getId(parent));
+};
+
+function tocObj(str, options = {}) {
+  options = Object.assign({
+    min_depth: 0,
```
Comment (noraj) [inline lib/toc_obj.js]:
```diff
-     min_depth: 0, 
+     min_depth: 1, 
```

File: lib/toc_obj.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    max_depth: 6
+  }, options);
+
+  const headingsSelector = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].slice(options.min_depth, options.max_depth).join(',');
```
Comment (noraj) [inline lib/toc_obj.js]:
```diff
-slice(options.min_depth, options.max_depth)
+slice(options.min_depth - 1, options.max_depth)
```
see my PR https://github.com/hexojs/hexo/pull/3929/files

File: lib/toc_obj.js (hunk 2 above)
Comment (noraj) [reply lib/toc_obj.js]: :heavy_check_mark:

File: lib/toc_obj.js (hunk 3 above)
Comment (noraj) [reply lib/toc_obj.js]: :heavy_check_mark:

File: README.md (hunk 1 above)
Comment (noraj) [reply README.md]: You forgot to change this one.

**Pull request merged**